tests/templates/test_subroutines/test_controlled_sequence.py

```diff
@@ -0,0 +1,408 @@
+        @qml.compile()
+        def circuit(thetas):
+            qml.ControlledSequence(U(thetas), control=[2, 3])
+            return qml.state()
+
+        _ = circuit([1.0, 1.0])
+
+    def test_approx_time_base(self):
+        """Test that using ControlledSequence with ApproxTimeEvolution as a base in a
+        circuit can be decomposed to execute successfully"""
+        H = qml.Hamiltonian([1.0, 2.0], [qml.PauliZ(0), qml.PauliZ(1)])
+        dev = qml.device("default.qubit", wires=4)
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.ControlledSequence(qml.ApproxTimeEvolution(H, 1, 1), control=[2, 3])
+            return qml.state()
+
+        _ = circuit()
+
+    @pytest.mark.xfail(reason="not working yet")
+    def test_gradient_with_composite_op_base(self):
+        """Test executing and getting the gradient of a circuit with a
+        ControlledSequence based on a CompositeOp"""
+
+        @qml.prod
+        def U(thetas):
+            qml.RX(thetas[0], wires=0)
+            qml.RX(thetas[1], wires=1)
+
+        dev = qml.device("default.qubit", wires=4)
+
+        @qml.qnode(dev)
+        def circuit(thetas):
+            qml.ControlledSequence(U(thetas, wires=[0, 1]), control=[2, 3])
+            return qml.state()
+
+        thetas = np.array([1.0, 1.0], requires_grad=True)
+        _ = circuit(thetas)
+        _ = qml.grad(circuit)(thetas)
+
+
+class TestQPEResults:
+    """Tests phase estimation using the template"""
+
+    @pytest.mark.parametrize("phase", [7, np.pi, np.pi / 3, 2.3])
+    def test_phase_estimated_single_ops(self, phase):
+        """Tests that the QPE defined using ControlledSequence works correctly for a single operator"""
+        dev = qml.device("default.qubit")
+        unitary = qml.RX(phase, wires=[0])
+
+        estimates = []
+        wire_range = range(3, 11)
+
+        for wires in wire_range:
+            estimation_wires = range(1, wires - 1)
+
+            @qml.qnode(dev)
+            def circuit():
+                qml.Hadamard(wires=0)
+
+                for i in estimation_wires:
+                    qml.Hadamard(wires=i)
+
+                qml.ControlledSequence(unitary, control=estimation_wires)
+
+                qml.adjoint(qml.QFT)(wires=estimation_wires)
+
+                return qml.probs(wires=estimation_wires)
+
+            res = circuit()
+
+            estimate = np.argmax(res) / 2 ** (wires - 2)
+            # Need to rescale phase due to convention of RX gate
+            estimate = 4 * np.pi * (1 - estimate)
+            estimates.append(estimate)
+
+        # Check that the error is monotonically decreasing
+        for i in range(len(estimates) - 1):
+            err1 = np.abs(estimates[i] - phase)
+            err2 = np.abs(estimates[i + 1] - phase)
+            assert err1 >= err2
+
+        # This is a large error, but we'd need to push the qubit number up more to get it lower
+        assert np.allclose(estimates[-1], phase, rtol=2e-2)
```

Comment:
Are these useful tests?

Reply:
I think it is great way to test is it by checking it is  equivalent to the QPE. I have changed the test to make it clearer using `qml.QuantumPhaseEstimation`.